mirascope/core/mistral/message.py

```diff
@@ -0,0 +1,40 @@
+from typing import Annotated, Any, Literal, TypeAlias
+
+from mistralai.models import AssistantMessage, SystemMessage, ToolMessage, UserMessage
+from pydantic import Discriminator, Tag, TypeAdapter
+
+
+def is_chat_message(v: Any) -> bool:  # noqa: ANN401
+    return isinstance(v, SystemMessage | UserMessage | AssistantMessage | ToolMessage)
+
+
+def chat_message_discriminator(v) -> Literal["system", "user", "assistant", "tool"]:  # noqa: ANN001
+    """Discriminator the the ChatMessage union.
+
+    Unfortunately, the 'role' field in the sub-types of the union is Optional, which is not allowed for a discriminator field.
+    See ?? for more details.
+
+    By using a simple mapping function we can remove the optionality of the type, at the cost of not being able to to have a
+    missing role.
+    """
+    role = v.get("role") if isinstance(v, dict) else v.role
+    if role is not None:
+        return role
+    raise ValueError("role cannot be None")
+
+
+ChatMessage: TypeAlias = Annotated[
+    Annotated[SystemMessage, Tag("system")]
+    | Annotated[UserMessage, Tag("user")]
+    | Annotated[AssistantMessage, Tag("assistant")]
+    | Annotated[ToolMessage, Tag("tool")],
+    Discriminator(chat_message_discriminator),
+]
+
+
+def make_chat_message(**kwargs: Any) -> ChatMessage:  # noqa: ANN401
```

Comment:
this feels like an internal method? I feel like we should move this into a `_make_chat_message.py` file in `_utils` to retain more control over changes to it as necessary (rather than making it public).

My other question is how necessary this function is at all? Can we not just internally use their types and check the role and in the message conversion just create the correct role?

We could then put this as a private utility of the conversion utility and likely drop the whole discriminator thing?

Reply:
Moving it to internal makes sense.

Hmm I'm not sure what the broader suggestion here is. Are you suggesting we could get rid of this function and then anywhere we need to construct a "ChatMessage" we would do it inline? Or are you maybe suggestioning a different implementation of this function?

Reply:
Yeah I'm suggesting we could get rid of this function and just construct the appropriate messages inline.

However, since they offer two methods for writing messages (e.g. `UserMessage` and `UserMessageTypedDict` we'll need to handle both of these cases, in which case we could have `_convert_message_to_base_message_param` as an internal helper function in the `_convert_message_params.py` utility file.

Does this make sense?